### Final Project Requirements/notes: https://docs.google.com/document/d/1mwYbYJHkB7kpx4tNflKh54jN9_oOscw3p4k5fsmn3bc/edit

In [ ]:
# look into contains in pandas, and pandas mask

### Link with all Data: https://www.sec.gov/dera/data/financial-statement-and-notes-data-set.html
- using NUM file only for now (data set of all numeric XBRL facts presented on the primary financial statements)

In [72]:
import pandas as pd
quarter_numbers = pd.read_table('2012q1_notes/num.tsv', encoding ='latin1')

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [40]:
quarter_numbers.head()

adsh                                     tag  \
8   0001376231-12-000002      RetainedEarningsAccumulatedDeficit   
9   0001376231-12-000002      RetainedEarningsAccumulatedDeficit   
13  0001193125-12-080059            DueFromRelatedPartiesCurrent   
14  0001193125-12-080059            DueFromRelatedPartiesCurrent   
15  0001193125-12-080059  IncreaseDecreaseInOtherAccountsPayable   

         version     ddate  qtrs  uom        dimh  iprx       value  footlen  \
8   us-gaap/2011  20101231     0  USD  0x00000000   0.0   -107439.0        0   
9   us-gaap/2011  20111231     0  USD  0x00000000   0.0   -112835.0        0   
13  us-gaap/2011  20111231     0  USD  0x00000000   0.0   5903000.0        0   
14  us-gaap/2011  20101231     0  USD  0x00000000   0.0   4921000.0        0   
15  us-gaap/2011  20111231     4  USD  0x00000000   0.0  14577000.0        0   

    dimn      durp  datp  dcml  
8      0  0.000000   0.0   0.0  
9      0  0.000000   0.0   0.0  
13     0  0.000000   0.0  -3.0  
14     0  0.000000   0.0  -3.0  
15     0  0.010959   0.0  -3.0

In [73]:
quarter_numbers = quarter_numbers.drop(columns=['footnote','coreg'])
quarter_numbers = quarter_numbers.dropna()

In [5]:
import numpy as np
tags = quarter_numbers.tag.unique()

In [6]:
tags.shape

(190458,)

In [8]:
dates = quarter_numbers.ddate.unique()
len(dates)

355

In [74]:
nums_2011 = quarter_numbers.loc[quarter_numbers['ddate'] == 20111231]
len(nums_2011)

1719181

In [11]:
tags = nums_2011.tag.unique()
len(tags)

142652

In [14]:
len(nums_2011.adsh.unique())

6280

In [15]:
10/6280

0.0015923566878980893

In [76]:
nums_2011 = nums_2011[['adsh','tag']]
nums_2011.shape

(1719181, 2)

In [77]:
temp = nums_2011[0:10]

In [68]:
# for i,row in enumerate(nums_2011[0:10]):

#     if 'Assets' in row[1]:
#         nums_2011[i][1] = 'AccountsReceivable'


In [78]:
temp

adsh                                                tag
9   0001376231-12-000002                 RetainedEarningsAccumulatedDeficit
13  0001193125-12-080059                       DueFromRelatedPartiesCurrent
15  0001193125-12-080059             IncreaseDecreaseInOtherAccountsPayable
20  0001193125-12-080059  OtherComprehensiveIncomeDerivativesQualifyingA...
21  0001193125-12-080059  NetCashProvidedByUsedInFinancingActivitiesCont...
24  0001193125-12-080059                       ProceedsFromIncomeTaxRefunds
29  0001193125-12-080059                  LoansToParentReclassifiedToEquity
31  0001144204-12-016089                                             Assets
33  0001144204-12-016089                            CommonStockSharesIssued
35  0001144204-12-016089                      GainLossOnSaleOfSecuritiesNet

In [79]:
#lowercases all
temp['tag'] = temp['tag'].str.lower()

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [60]:
# temp.loc[temp['tag'].str.contains("accounts|assets")] = 'asset'

/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:189: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [81]:
assets = temp.loc[temp['tag'].str.contains("accounts|assets")]['tag']
assets

15    increasedecreaseinotheraccountspayable
31                                    assets
Name: tag, dtype: object

In [87]:
temp.loc[temp['tag'].str.contains("accounts|assets")]['tag']

15    increasedecreaseinotheraccountspayable
31                                    assets
Name: tag, dtype: object

In [95]:
x = temp.loc[temp['tag'].str.contains("accounts|assets")]
x.tag = 'assets'
x

adsh     tag
15  0001193125-12-080059  assets
31  0001144204-12-016089  assets

In [100]:
temp.loc[temp['tag'].str.contains("accounts|assets")].index

Int64Index([15, 31], dtype='int64')

In [104]:
temp = temp.set_value(temp.loc[temp['tag'].str.contains("accounts|assets")].index, 'tag', 'assets')

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  """Entry point for launching an IPython kernel.


In [105]:
temp

adsh                                                tag
9   0001376231-12-000002                 retainedearningsaccumulateddeficit
13  0001193125-12-080059                       duefromrelatedpartiescurrent
15  0001193125-12-080059                                             assets
20  0001193125-12-080059  othercomprehensiveincomederivativesqualifyinga...
21  0001193125-12-080059  netcashprovidedbyusedinfinancingactivitiescont...
24  0001193125-12-080059                       proceedsfromincometaxrefunds
29  0001193125-12-080059                  loanstoparentreclassifiedtoequity
31  0001144204-12-016089                                             assets
33  0001144204-12-016089                            commonstocksharesissued
35  0001144204-12-016089                      gainlossonsaleofsecuritiesnet

In [75]:
temp2 = nums_2011[0:10]
temp2

adsh                                                tag  \
9   0001376231-12-000002                 RetainedEarningsAccumulatedDeficit   
13  0001193125-12-080059                       DueFromRelatedPartiesCurrent   
15  0001193125-12-080059             IncreaseDecreaseInOtherAccountsPayable   
20  0001193125-12-080059  OtherComprehensiveIncomeDerivativesQualifyingA...   
21  0001193125-12-080059  NetCashProvidedByUsedInFinancingActivitiesCont...   
24  0001193125-12-080059                       ProceedsFromIncomeTaxRefunds   
29  0001193125-12-080059                  LoansToParentReclassifiedToEquity   
31  0001144204-12-016089                                             Assets   
33  0001144204-12-016089                            CommonStockSharesIssued   
35  0001144204-12-016089                      GainLossOnSaleOfSecuritiesNet   

                 version     ddate  qtrs     uom        dimh  iprx  \
9           us-gaap/2011  20111231     0     USD  0x00000000   0.0   
13          us-gaap/2011  20111231     0     USD  0x00000000   0.0   
15          us-gaap/2011  20111231     4     USD  0x00000000   0.0   
20          us-gaap/2011  20111231     4     USD  0x00000000   0.0   
21          us-gaap/2011  20111231     4     USD  0x00000000   0.0   
24          us-gaap/2011  20111231     4     USD  0x00000000   0.0   
29  0001193125-12-080059  20111231     4     USD  0x00000000   0.0   
31          us-gaap/2011  20111231     0     USD  0x00000000   0.0   
33          us-gaap/2011  20111231     0  shares  0x00000000   0.0   
35          us-gaap/2011  20111231     4     USD  0x00000000   0.0   

          value  footlen  dimn      durp  datp  dcml  
9     -112835.0        0     0  0.000000   0.0   0.0  
13    5903000.0        0     0  0.000000   0.0  -3.0  
15   14577000.0        0     0  0.010959   0.0  -3.0  
20    -519000.0        0     0  0.010959   0.0  -3.0  
21  -40009000.0        0     0  0.010959   0.0  -3.0  
24          0.0        0     0  0.010959   0.0  -3.0  
29          0.0        0     0  0.010959   0.0  -3.0  
31  469486590.0        0     0  0.000000   0.0   0.0  
33    5474437.0        0     0  0.000000   0.0   0.0  
35     -43952.0        0     0  0.010959   0.0   0.0

In [24]:
tags

array(['RetainedEarningsAccumulatedDeficit',
       'DueFromRelatedPartiesCurrent',
       'IncreaseDecreaseInOtherAccountsPayable', ...,
       'LeasedSpaceSquareFootage',
       'NoncontrollingInterestsInMidamericanEnergySubsidiary',
       'MidamericanFundingDebtInstrumentInterestRate'], dtype=object)

In [ ]:
a= tags_list
for n, i in enumerate(a):
    if i.lower().startswith(('accountsreceivable','accountreceivable')):
        a[n] = 'AccountsReceivable'
len(set(tags_list))

In [9]:
type(tags)

numpy.ndarray

In [16]:
tags_list = tags.tolist()

In [17]:
sorted(tags_list)

['A001AdditionalPaidinCapitalWarrants',
 'A001SharesToBeIssued',
 'A010OrganizationCostsWriteoff',
 'A011WeightedAverageNumberOfSharesOutstandingBasicAndDiluted',
 'A012StockIssuedConversionOfNotePayableValue',
 'A013StockIssuedConversionOfNotePayableShares',
 'A014SharesIssuedForServicesValue',
 'A015SharesIssuedForServicesShares',
 'A02WeightedAverageSharesOutstanding',
 'A10YearTerm',
 'A2000EquityIncentivePlanAwardsGrantedInPeriod',
 'A2000EquityIncentivePlanRemainingVestingPeriod',
 'A2007RetiredStockRepurchaseProgramNumberOfSharesAuthorizedToBeRepurchased',
 'A2008GmAgreementCashReceivedIn2008',
 'A2008GmAgreementCashReceivedIn2009',
 'A2008GmAgreementFairValueOfLiabilityAssumedByGmForPostretirementCostSharing',
 'A2008GmAgreementSharesIssuedToGmThroughExerciseOfWarrants',
 'A2008GmAgreementTotalDeferredRevenueRecognized',
 'A2009AmendmentToStockholderRightsPlanPercentageAtWhichPersonBecomesAcquiringPersonAmended',
 'A2009AmendmentToStockholderRightsPlanPercentageAtWhichPersonBec

In [19]:
len(set(tags_list))

190458

In [25]:
result = filter(lambda x: x.startswith('AccountsReceivable'), tags_list)

In [26]:
result

In [27]:
result2 = [i for i in tags_list if i.startswith('AccountsReceivable')]

In [42]:
a= tags_list
for n, i in enumerate(a):
    if i.lower().startswith(('accountsreceivable','accountreceivable')):
        a[n] = 'AccountsReceivable'
len(set(tags_list))

189990

In [43]:
a= tags_list
for n, i in enumerate(a):
    if (i.lower()).startswith(('accountspayable', 'accountpayable')):
        a[n] = 'accountspayable'
len(set(tags_list))

189988

In [40]:
a= tags_list
for n, i in enumerate(a):
    if (i.lower()).startswith('abandoned'):
        a[n] = 'abandonedcosts'
len(set(tags_list))

190001

In [49]:
result3 = [i for i in tags_list if ('Receivable') in i]

In [50]:
len(result3)

6233

In [51]:
sorted(set(tags_list))

['A001AdditionalPaidinCapitalWarrants',
 'A001SharesToBeIssued',
 'A010OrganizationCostsWriteoff',
 'A011WeightedAverageNumberOfSharesOutstandingBasicAndDiluted',
 'A012StockIssuedConversionOfNotePayableValue',
 'A013StockIssuedConversionOfNotePayableShares',
 'A014SharesIssuedForServicesValue',
 'A015SharesIssuedForServicesShares',
 'A02WeightedAverageSharesOutstanding',
 'A10YearTerm',
 'A2000EquityIncentivePlanAwardsGrantedInPeriod',
 'A2000EquityIncentivePlanRemainingVestingPeriod',
 'A2007RetiredStockRepurchaseProgramNumberOfSharesAuthorizedToBeRepurchased',
 'A2008GmAgreementCashReceivedIn2008',
 'A2008GmAgreementCashReceivedIn2009',
 'A2008GmAgreementFairValueOfLiabilityAssumedByGmForPostretirementCostSharing',
 'A2008GmAgreementSharesIssuedToGmThroughExerciseOfWarrants',
 'A2008GmAgreementTotalDeferredRevenueRecognized',
 'A2009AmendmentToStockholderRightsPlanPercentageAtWhichPersonBecomesAcquiringPersonAmended',
 'A2009AmendmentToStockholderRightsPlanPercentageAtWhichPersonBec

In [ ]:
a= tags_list
for n, i in enumerate(a):
    if i.startswith('AccountsReceivable'):
        a[n] = 'AccountsReceivable'
len(set(tags_list))

In [29]:
len(result2)

253

In [6]:
quarter_numbers.tail()

adsh                         tag  \
4626815  0000928576-12-000003       UnsecuredLongTermDebt   
4626816  0000928576-12-000003  DebtInstrumentInterestRate   
4626817  0000928576-12-000003  DebtInstrumentInterestRate   
4626818  0000928576-12-000003       UnsecuredLongTermDebt   
4626819  0000928576-12-000003       UnsecuredLongTermDebt   

                      version     ddate  qtrs   uom  \
4626815          us-gaap/2011  20101231     0   USD   
4626816  0000928576-12-000003  20101231     4  pure   
4626817  0000928576-12-000003  20111231     4  pure   
4626818          us-gaap/2011  20101231     0   USD   
4626819          us-gaap/2011  20111231     0   USD   

                                       dimh  iprx         value footnote  \
4626815  0xf704973033dd1a254c01201245aa7d2c   0.0  5.700000e+07      NaN   
4626816  0xfb7b914638117c8664b80fa1b5ae9233   0.0  5.130000e-02      NaN   
4626817  0xfb7b914638117c8664b80fa1b5ae9233   0.0  5.130000e-02      NaN   
4626818  0xfb7b914638117c8664b80fa1b5ae9233   0.0  2.750000e+08      NaN   
4626819  0xfb7b914638117c8664b80fa1b5ae9233   0.0  2.750000e+08      NaN   

         footlen  dimn coreg      durp  datp     dcml  
4626815        0     2   NaN  0.000000   0.0     -6.0  
4626816        0     2   NaN  0.010959   0.0  32767.0  
4626817        0     2   NaN  0.010959   0.0  32767.0  
4626818        0     2   NaN  0.000000   0.0     -6.0  
4626819        0     2   NaN  0.000000   0.0     -6.0

In [2]:
#get details on all columns
# quarter_numbers.describe(include='all')

### From "Financial Statement and Notes Data Sets" Readme:
These fields comprise a unique compound key:

1) **adsh - EDGAR accession number**: a unique identifier assigned automatically to an accepted submission by the EDGAR Filer System; The first set of numbers (0001193125) is the CIK of the entity submitting the filing. The next 2 numbers (18) represent the year. The last series of numbers represent a sequential count of submitted filings from that CIK. The count is usually, but not always, reset to 0 at the start of each calendar year.

2) **tag** - tag used by the filer 
- **TODO**: may have to separate out first word from tag to identify broader groups such as revenue

3) **version** – if a standard tag, the taxonomy of origin, otherwise equal to adsh.

4) **ddate** - period end date

5) **qtrs** - duration in number of quarters

6) **uom** - unit of measure

7) **dimh** - 16-byte dimensional qualifier

8) **iprx** - a sequential integer used to distinguish otherwise identical facts

9) **coreg** - If specified, indicates a specific co-registrant, the parent company, or other entity (e.g., guarantor).  NULL indicates the consolidated entity.  Note that this value is a function of the dimension segments.

10) **durp** - The difference between the reported fact duration and the quarter duration (qtrs), expressed as a fraction of 1.  For example, a fact with duration of 120 days rounded to a 91-day quarter has a durp value of 29/91 = +0.3187.

11) **datp** - The difference between the reported fact date and the month-end rounded date (ddate), expressed as a fraction of 1.  For example, a fact reported for 29/Dec, with ddate rounded to 31/Dec, has a datp value of minus 2/31 = -0.0645.
 
12) **dcml** - The value of the fact "decimals" attribute, with INF represented by 32767.

In [3]:
# quarter_numbers.isnull().sum()

adsh              0
tag               0
version           0
ddate             0
qtrs              0
uom               0
dimh              0
iprx         333394
value         82992
footnote    4609706
footlen           0
dimn              0
coreg       4364002
durp         333394
datp         333394
dcml         333394
dtype: int64

#### A lot of null values for footnotes and coregistrants (majority of rows); will remove these columns for now

In [7]:
quarter_numbers = quarter_numbers.drop(columns=['footnote','coreg'])

In [8]:
quarter_numbers.isnull().sum() 

adsh            0
tag             0
version         0
ddate           0
qtrs            0
uom             0
dimh            0
iprx       333394
value       82992
footlen         0
dimn            0
durp       333394
datp       333394
dcml       333394
dtype: int64

In [6]:
# q114numbers = q114numbers.drop(columns=['iprx','durp','datp','dcml'])

#### Didn't get rid of blank Value fields, so just get rid of those since there will still be enough data points to analyze

In [9]:
quarter_numbers = quarter_numbers.dropna()

In [10]:
quarter_numbers.describe(include='all') #still have 4million+ data points

adsh                 tag       version         ddate  \
count                4214049             4214049       4214049  4.214049e+06   
unique                  7335              190458          6820           NaN   
top     0001140361-12-007480  StockholdersEquity  us-gaap/2011           NaN   
freq                   24832               88047       3486735           NaN   
mean                     NaN                 NaN           NaN  2.010420e+07   
std                      NaN                 NaN           NaN  1.121092e+04   
min                      NaN                 NaN           NaN  1.927023e+07   
25%                      NaN                 NaN           NaN  2.010123e+07   
50%                      NaN                 NaN           NaN  2.011023e+07   
75%                      NaN                 NaN           NaN  2.011123e+07   
max                      NaN                 NaN           NaN  2.199123e+07   

                qtrs      uom        dimh          iprx         value  \
count   4.214049e+06  4214049     4214049  4.214049e+06  4.214049e+06   
unique           NaN     1730      191868           NaN           NaN   
top              NaN      USD  0x00000000           NaN           NaN   
freq             NaN  3670481     2180713           NaN           NaN   
mean    2.258344e+00      NaN         NaN  3.502570e-04  3.774747e+12   
std     4.071756e+00      NaN         NaN  1.908854e-02  4.852022e+15   
min     0.000000e+00      NaN         NaN  0.000000e+00 -2.709478e+12   
25%     0.000000e+00      NaN         NaN  0.000000e+00  8.500000e-01   
50%     2.000000e+00      NaN         NaN  0.000000e+00  1.966000e+06   
75%     4.000000e+00      NaN         NaN  0.000000e+00  4.163000e+07   
max     1.201000e+03      NaN         NaN  5.000000e+00  7.399938e+18   

             footlen          dimn          durp          datp          dcml  
count   4.214049e+06  4.214049e+06  4.214049e+06  4.214049e+06  4.214049e+06  
unique           NaN           NaN           NaN           NaN           NaN  
top              NaN           NaN           NaN           NaN           NaN  
freq             NaN           NaN           NaN           NaN           NaN  
mean    7.252426e-01  6.416748e-01  4.414548e-03  7.746849e-02  2.261718e+03  
std     1.893733e+01  8.074701e-01  4.421040e-02  1.243504e+00  8.311397e+03  
min     0.000000e+00  0.000000e+00 -4.986305e-01 -1.500000e+01 -1.200000e+01  
25%     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00 -3.000000e+00  
50%     0.000000e+00  0.000000e+00  2.740025e-03  0.000000e+00 -3.000000e+00  
75%     0.000000e+00  1.000000e+00  1.095891e-02  0.000000e+00  0.000000e+00  
max     5.101000e+03  7.000000e+00  4.986310e-01  1.500000e+01  3.276700e+04

In [11]:
quarter_numbers.head()

adsh                                     tag  \
8   0001376231-12-000002      RetainedEarningsAccumulatedDeficit   
9   0001376231-12-000002      RetainedEarningsAccumulatedDeficit   
13  0001193125-12-080059            DueFromRelatedPartiesCurrent   
14  0001193125-12-080059            DueFromRelatedPartiesCurrent   
15  0001193125-12-080059  IncreaseDecreaseInOtherAccountsPayable   

         version     ddate  qtrs  uom        dimh  iprx       value  footlen  \
8   us-gaap/2011  20101231     0  USD  0x00000000   0.0   -107439.0        0   
9   us-gaap/2011  20111231     0  USD  0x00000000   0.0   -112835.0        0   
13  us-gaap/2011  20111231     0  USD  0x00000000   0.0   5903000.0        0   
14  us-gaap/2011  20101231     0  USD  0x00000000   0.0   4921000.0        0   
15  us-gaap/2011  20111231     4  USD  0x00000000   0.0  14577000.0        0   

    dimn      durp  datp  dcml  
8      0  0.000000   0.0   0.0  
9      0  0.000000   0.0   0.0  
13     0  0.000000   0.0  -3.0  
14     0  0.000000   0.0  -3.0  
15     0  0.010959   0.0  -3.0

In [10]:
quarter_numbers.columns

Index(['adsh', 'tag', 'version', 'ddate', 'qtrs', 'uom', 'dimh', 'iprx',
       'value', 'footlen', 'dimn', 'durp', 'datp', 'dcml'],
      dtype='object')

In [11]:
#Drop some columns to see if can reindex with the tags

In [12]:
quarter_numbers = quarter_numbers.drop(columns=['version','qtrs','uom', 'dimh', 'iprx',
       'footlen', 'dimn', 'durp', 'datp', 'dcml'])

In [13]:
# quarter_numbers = quarter_numbers.set_index(quarter_numbers.adsh)
# quarter_numbers.head()

In [14]:
# quarter_numbers = quarter_numbers.drop(columns=['adsh'])
# quarter_numbers

In [15]:
# stacking_test = quarter_numbers.stack().to_frame()

In [16]:
# stacking_test

In [135]:
# explore = quarter_numbers.loc[quarter_numbers['adsh'] == '0001376231-12-000002']
# explore.head()

adsh                                                tag  \
8   0001376231-12-000002                 RetainedEarningsAccumulatedDeficit   
9   0001376231-12-000002                 RetainedEarningsAccumulatedDeficit   
76  0001376231-12-000002  IncreaseDecreaseInAccountsPayableAndAccruedLia...   
77  0001376231-12-000002  IncreaseDecreaseInAccountsPayableAndAccruedLia...   
78  0001376231-12-000002  IncreaseDecreaseInAccountsPayableAndAccruedLia...   

       ddate     value                                          tag_ddate  
8   20101231 -107439.0         RetainedEarningsAccumulatedDeficit20101231  
9   20111231 -112835.0         RetainedEarningsAccumulatedDeficit20111231  
76  20111231   -4536.0  IncreaseDecreaseInAccountsPayableAndAccruedLia...  
77  20101231    7871.0  IncreaseDecreaseInAccountsPayableAndAccruedLia...  
78  20111231   16027.0  IncreaseDecreaseInAccountsPayableAndAccruedLia...

In [136]:
# len(explore)

151

In [137]:
# explore = explore.set_index(explore.adsh)
# len(explore)

151

In [142]:
# explore2 = quarter_numbers.loc[quarter_numbers['adsh'] == '0001376231-12-000002']
# explore2.head()

adsh                                                tag  \
8   0001376231-12-000002                 RetainedEarningsAccumulatedDeficit   
9   0001376231-12-000002                 RetainedEarningsAccumulatedDeficit   
76  0001376231-12-000002  IncreaseDecreaseInAccountsPayableAndAccruedLia...   
77  0001376231-12-000002  IncreaseDecreaseInAccountsPayableAndAccruedLia...   
78  0001376231-12-000002  IncreaseDecreaseInAccountsPayableAndAccruedLia...   

       ddate     value                                          tag_ddate  
8   20101231 -107439.0         RetainedEarningsAccumulatedDeficit20101231  
9   20111231 -112835.0         RetainedEarningsAccumulatedDeficit20111231  
76  20111231   -4536.0  IncreaseDecreaseInAccountsPayableAndAccruedLia...  
77  20101231    7871.0  IncreaseDecreaseInAccountsPayableAndAccruedLia...  
78  20111231   16027.0  IncreaseDecreaseInAccountsPayableAndAccruedLia...

In [144]:
# explore2 = pd.pivot_table(explore2,index=["adsh","ddate"], columns= 'tag', values= 'value')
# explore2.head()

tag                            AccountsPayableAndOtherAccruedLiabilities  \
adsh                 ddate                                                 
0001376231-12-000002 20041231                                        NaN   
                     20051231                                        NaN   
                     20061231                                        NaN   
                     20071231                                        NaN   
                     20081231                                        NaN   

tag                            AccumulatedOtherComprehensiveIncomeLossNetOfTax  \
adsh                 ddate                                                       
0001376231-12-000002 20041231                                              NaN   
                     20051231                                              NaN   
                     20061231                                              NaN   
                     20071231                                              NaN   
                     20081231                                              NaN   

tag                            AdditionalPaidInCapital  Assets  \
adsh                 ddate                                       
0001376231-12-000002 20041231                      NaN     NaN   
                     20051231                      NaN     NaN   
                     20061231                      NaN     NaN   
                     20071231                      NaN     NaN   
                     20081231                      NaN     NaN   

tag                            CashAndCashEquivalentsAtCarryingValue  \
adsh                 ddate                                             
0001376231-12-000002 20041231                                    NaN   
                     20051231                                    NaN   
                     20061231                                    NaN   
                     20071231                                    NaN   
                     20081231                                    NaN   

tag                            DepreciationDepletionAndAmortization  \
adsh                 ddate                                            
0001376231-12-000002 20041231                                   NaN   
                     20051231                                   NaN   
                     20061231                                   NaN   
                     20071231                                   NaN   
                     20081231                                   NaN   

tag                            DueToRelatedPartiesCurrentAndNoncurrent  \
adsh                 ddate                                               
0001376231-12-000002 20041231                                      NaN   
                     20051231                                      NaN   
                     20061231                                      NaN   
                     20071231                                      NaN   
                     20081231                                      NaN   

tag                            EmployeeRelatedLiabilitiesCurrentAndNoncurrent  \
adsh                 ddate                                                      
0001376231-12-000002 20041231                                             NaN   
                     20051231                                             NaN   
                     20061231                                             NaN   
                     20071231                                             NaN   
                     20081231                                             NaN   

tag                            EntityCommonStockSharesOutstanding  \
adsh                 ddate                                          
0001376231-12-000002 20041231                                 NaN   
                     20051231                                 NaN   
                     20061231                                 N

In [145]:
# explore2

tag                            AccountsPayableAndOtherAccruedLiabilities  \
adsh                 ddate                                                 
0001376231-12-000002 20041231                                        NaN   
                     20051231                                        NaN   
                     20061231                                        NaN   
                     20071231                                        NaN   
                     20081231                                        NaN   
                     20091231                                        NaN   
                     20101231                                     7871.0   
                     20110630                                        NaN   
                     20111231                                     3125.0   
                     20120331                                        NaN   

tag                            AccumulatedOtherComprehensiveIncomeLossNetOfTax  \
adsh                 ddate                                                       
0001376231-12-000002 20041231                                              NaN   
                     20051231                                              NaN   
                     20061231                                              NaN   
                     20071231                                              NaN   
                     20081231                                              NaN   
                     20091231                                              NaN   
                     20101231                                          -1773.0   
                     20110630                                              NaN   
                     20111231                                          -2381.0   
                     20120331                                              NaN   

tag                            AdditionalPaidInCapital  Assets  \
adsh                 ddate                                       
0001376231-12-000002 20041231                      NaN     NaN   
                     20051231                      NaN     NaN   
                     20061231                      NaN     NaN   
                     20071231                      NaN     NaN   
                     20081231                      NaN     NaN   
                     20091231                      NaN     NaN   
                     20101231                   1772.0     0.0   
                     20110630                      NaN     NaN   
                     20111231                   1772.0   105.0   
                     20120331                      NaN     NaN   

tag                            CashAndCashEquivalentsAtCarryingValue  \
adsh                 ddate                                             
0001376231-12-000002 20041231                                    NaN   
                     20051231                                    NaN   
                     20061231                                    NaN   
                     20071231                                    NaN   
                     20081231                                    NaN   
                     20091231                                    NaN   
                     20101231                                    0.0   
                     20110630                                    NaN   
                     20111231                                    0.0   
                     20120331                                    NaN   

tag                            DepreciationDepletionAndAmortization  \
adsh                 ddate                                            
0001376231-12-000002 20041231                                   NaN   
                     20051231                                   NaN   
                     20061231                                   NaN   
                     20071231                                   NaN   
                  

In [146]:
# explore3 = quarter_numbers.loc[quarter_numbers['adsh'] == '0001376231-12-000002']
# explore3 = pd.pivot_table(explore3,index=["adsh"], columns= 'tag_ddate', values= 'value')

In [147]:
# explore3

tag_ddate             AccountsPayableAndOtherAccruedLiabilities20101231  \
adsh                                                                      
0001376231-12-000002                                             7871.0   

tag_ddate             AccountsPayableAndOtherAccruedLiabilities20111231  \
adsh                                                                      
0001376231-12-000002                                             3125.0   

tag_ddate             AccumulatedOtherComprehensiveIncomeLossNetOfTax20101231  \
adsh                                                                            
0001376231-12-000002                                            -1773.0         

tag_ddate             AccumulatedOtherComprehensiveIncomeLossNetOfTax20111231  \
adsh                                                                            
0001376231-12-000002                                            -2381.0         

tag_ddate             AdditionalPaidInCapital20101231  \
adsh                                                    
0001376231-12-000002                           1772.0   

tag_ddate             AdditionalPaidInCapital20111231  Assets20101231  \
adsh                                                                    
0001376231-12-000002                           1772.0             0.0   

tag_ddate             Assets20111231  \
adsh                                   
0001376231-12-000002           105.0   

tag_ddate             CashAndCashEquivalentsAtCarryingValue20101231  \
adsh                                                                  
0001376231-12-000002                                            0.0   

tag_ddate             CashAndCashEquivalentsAtCarryingValue20111231  \
adsh                                                                  
0001376231-12-000002                                            0.0   

tag_ddate                                                             ...                                                   \
adsh                                                                  ...                                                    
0001376231-12-000002                                                  ...                                                    

tag_ddate             ProceedsFromMinorityShareholders20111231  \
adsh                                                             
0001376231-12-000002                                   47325.0   

tag_ddate             ProfitLoss20101231  ProfitLoss20111231  \
adsh                                                           
0001376231-12-000002             -8471.0            -59115.5   

tag_ddate             ResearchAndDevelopmentExpense20111231  \
adsh                                                          
0001376231-12-000002                                 3017.0   

tag_ddate             RetainedEarningsAccumulatedDeficit20101231  \
adsh                                                               
0001376231-12-000002                                   -107439.0   

tag_ddate             RetainedEarningsAccumulatedDeficit20111231  \
adsh                                                               
0001376231-12-000002                                   -112835.0   

tag_ddate             Revenues20111231  ShareBasedCompensation20111231  \
adsh                                                                     
0001376231-12-000002            6303.0                         41643.0   

tag_ddate             WeightedAverageNumberOfLimitedPartnershipAndGeneralPartnershipUnitOutstandingBasicAndDiluted20101231  \
adsh                                                                                                                         
0001376231-12-000002                                          2625425.0                                                      

tag_ddate             WeightedAverageNumberOfLimitedPartnershipAndGeneralPartnershipUnitOutstandingBasicAndDiluted20111231  
adsh                 

In [13]:
explore4 = quarter_numbers
explore4.head()

adsh                                     tag     ddate  \
8   0001376231-12-000002      RetainedEarningsAccumulatedDeficit  20101231   
9   0001376231-12-000002      RetainedEarningsAccumulatedDeficit  20111231   
13  0001193125-12-080059            DueFromRelatedPartiesCurrent  20111231   
14  0001193125-12-080059            DueFromRelatedPartiesCurrent  20101231   
15  0001193125-12-080059  IncreaseDecreaseInOtherAccountsPayable  20111231   

         value  
8    -107439.0  
9    -112835.0  
13   5903000.0  
14   4921000.0  
15  14577000.0

In [ ]:
explore4["tag_ddate"] = explore4["tag"].astype(str)+ explore4["ddate"].astype(str)
explore4 = explore4.drop(columns = ['ddate', 'tag'])

In [15]:
explore4 = pd.pivot_table(explore4,index=["adsh"], columns= ['tag','ddate'], values= 'value',fill_value=0) #takes 10 min to run

In [16]:
explore4

tag                  A001AdditionalPaidinCapitalWarrants           \
ddate                                           20110531 20111130   
adsh                                                                
0000002178-12-000010                                 NaN      NaN   
0000003453-12-000026                                 NaN      NaN   
0000003499-12-000005                                 NaN      NaN   
0000003570-12-000023                                 NaN      NaN   
0000004127-12-000010                                 NaN      NaN   
0000004457-12-000008                                 NaN      NaN   
0000004904-12-000013                                 NaN      NaN   
0000005513-12-000006                                 NaN      NaN   
0000006207-12-000002                                 NaN      NaN   
0000006885-12-000026                                 NaN      NaN   
0000007039-12-000003                                 NaN      NaN   
0000007084-12-000010                                 NaN      NaN   
0000007332-12-000003                                 NaN      NaN   
0000008328-12-000022                                 NaN      NaN   
0000008670-12-000002                                 NaN      NaN   
0000008858-12-000004                                 NaN      NaN   
0000008868-12-000010                                 NaN      NaN   
0000009346-12-000005                                 NaN      NaN   
0000009984-12-000036                                 NaN      NaN   
0000011544-12-000021                                 NaN      NaN   
0000012208-12-000008                                 NaN      NaN   
0000012400-12-000005                                 NaN      NaN   
0000012779-12-000014                                 NaN      NaN   
0000012779-12-000020                                 NaN      NaN   
0000013239-12-000016                                 NaN      NaN   
0000014195-12-000007                                 NaN      NaN   
0000014693-12-000024                                 NaN      NaN   
0000014707-12-000029                                 NaN      NaN   
0000014930-12-000028                                 NaN      NaN   
0000015511-12-000006                                 NaN      NaN   
...                                                  ...      ...   
0001525878-12-000017                                 NaN      NaN   
0001525878-12-000038                                 NaN      NaN   
0001525878-12-000039                                 NaN      NaN   
0001525878-12-000048                                 NaN      NaN   
0001525878-12-000078                                 NaN      NaN   
0001526508-12-000005                                 NaN      NaN   
0001528308-12-000005                                 NaN      NaN   
0001529290-12-000003                                 NaN      NaN   
0001529290-12-000028                                 NaN      NaN   
0001529293-12-000002                                 NaN      NaN   
0001529293-12-000004                                 NaN      NaN   
0001529293-12-000006                                 NaN      NaN   
0001530950-12-000071                                 NaN      NaN   
0001530981-12-000007                                 NaN      NaN   
0001530981-12-000015                                 NaN      NaN   
0001530981-12-000017                                 NaN      NaN   
0001530981-12-000020                                 NaN      NaN   
0001530981-12-000027                                 NaN      NaN   
0001531434-12-000006                                 NaN      NaN   
0001531618-12-000003                                 NaN      NaN   
0001532619-12-000013                                 NaN      NaN   
0001533455-12-000002                                 NaN      NaN   
0001533621-12-000021                                 NaN      NaN   
0001533621-12-000023                                 NaN      NaN  

In [ ]:
explore5 = explore4.fillna(0)

In [134]:
explore = explore.drop(columns = ['tag_ddate'])
explore.head()

adsh                                                tag  \
8   0001376231-12-000002                 RetainedEarningsAccumulatedDeficit   
9   0001376231-12-000002                 RetainedEarningsAccumulatedDeficit   
76  0001376231-12-000002  IncreaseDecreaseInAccountsPayableAndAccruedLia...   
77  0001376231-12-000002  IncreaseDecreaseInAccountsPayableAndAccruedLia...   
78  0001376231-12-000002  IncreaseDecreaseInAccountsPayableAndAccruedLia...   

       ddate     value  
8   20101231 -107439.0  
9   20111231 -112835.0  
76  20111231   -4536.0  
77  20101231    7871.0  
78  20111231   16027.0

In [131]:
explore = explore.set_index(['ddate'])
explore.head()

adsh  \
ddate                            
20101231  0001376231-12-000002   
20111231  0001376231-12-000002   
20111231  0001376231-12-000002   
20101231  0001376231-12-000002   
20111231  0001376231-12-000002   

                                                        tag     value  
ddate                                                                  
20101231                 RetainedEarningsAccumulatedDeficit -107439.0  
20111231                 RetainedEarningsAccumulatedDeficit -112835.0  
20111231  IncreaseDecreaseInAccountsPayableAndAccruedLia...   -4536.0  
20101231  IncreaseDecreaseInAccountsPayableAndAccruedLia...    7871.0  
20111231  IncreaseDecreaseInAccountsPayableAndAccruedLia...   16027.0

In [127]:
# index = explore[['adsh','tag']]

In [132]:
pivot = explore.pivot(index= 'adsh',  columns='tag', values='value')

ValueError: Index contains duplicate entries, cannot reshape

In [118]:
pivot = explore.pivot(index=['adsh',  columns='tag_ddate', values='value')

SyntaxError: invalid syntax (<ipython-input-118-f475a081f7f4>, line 1)

In [112]:
indexing = quarter_numbers[0:60]
indexing["tag_ddate"] = indexing["tag"].astype(str)+ indexing["ddate"].astype(str)
indexing = indexing.drop(columns = ['ddate', 'tag'])
pivot = indexing.pivot(index='adsh', columns='tag_ddate', values='value')
pivot.head()
# set_index(['adsh','tag','ddate'])
# indexing = indexing.drop(columns = ['ddate'])

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


tag_ddate             AccumulatedOtherComprehensiveIncomeLossNetOfTax20110930  \
adsh                                                                            
0001047469-12-000847                                                NaN         
0001144204-12-016089                                                NaN         
0001144204-12-017161                                                NaN         
0001193125-12-080059                                                NaN         
0001294154-12-000030                                        -14613000.0         

tag_ddate             AccumulatedOtherComprehensiveIncomeLossNetOfTax20111231  \
adsh                                                                            
0001047469-12-000847                                                NaN         
0001144204-12-016089                                                NaN         
0001144204-12-017161                                                NaN         
0001193125-12-080059                                                NaN         
0001294154-12-000030                                        -13504000.0         

tag_ddate             Assets20101231  Assets20111231  \
adsh                                                   
0001047469-12-000847    1.870052e+09    1.689668e+09   
0001144204-12-016089    4.544758e+08    4.694866e+08   
0001144204-12-017161             NaN             NaN   
0001193125-12-080059             NaN             NaN   
0001294154-12-000030             NaN             NaN   

tag_ddate             CommonStockSharesIssued20101231  \
adsh                                                    
0001047469-12-000847                              NaN   
0001144204-12-016089                        7225000.0   
0001144204-12-017161                              NaN   
0001193125-12-080059                              NaN   
0001294154-12-000030                              NaN   

tag_ddate             CommonStockSharesIssued20111231  \
adsh                                                    
0001047469-12-000847                              NaN   
0001144204-12-016089                        5474437.0   
0001144204-12-017161                              NaN   
0001193125-12-080059                              NaN   
0001294154-12-000030                              NaN   

tag_ddate             DeferredIncomeTaxExpenseBenefit20101231  \
adsh                                                            
0001047469-12-000847                                      NaN   
0001144204-12-016089                                      NaN   
0001144204-12-017161                                      NaN   
0001193125-12-080059                                      NaN   
0001294154-12-000030                                 422000.0   

tag_ddate             DeferredIncomeTaxExpenseBenefit20111231  \
adsh                                                            
0001047469-12-000847                                      NaN   
0001144204-12-016089                                      NaN   
0001144204-12-017161                                      NaN   
0001193125-12-080059                                      NaN   
0001294154-12-000030                                 154000.0   

tag_ddate             DistributionOfStockHeldInDeferredCompensationPlan20111231  \
adsh                                                                              
0001047469-12-000847                                           237000.0           
0001144204-12-016089                                                NaN           
0001144204-12-017161                                                NaN           
0001193125-12-080059                                                NaN           
0001294154-12-000030                                                NaN           

tag_ddate             DueFromRelatedPartiesCurrent20101231  \
adsh                                                         
0001047469-12-000847                                 

In [52]:
indexing = indexing.set_index(['adsh'])
indexing.head()

tag     ddate  \
adsh                                                                     
0001376231-12-000002      RetainedEarningsAccumulatedDeficit  20101231   
0001376231-12-000002      RetainedEarningsAccumulatedDeficit  20111231   
0001193125-12-080059            DueFromRelatedPartiesCurrent  20111231   
0001193125-12-080059            DueFromRelatedPartiesCurrent  20101231   
0001193125-12-080059  IncreaseDecreaseInOtherAccountsPayable  20111231   

                           value  
adsh                              
0001376231-12-000002   -107439.0  
0001376231-12-000002   -112835.0  
0001193125-12-080059   5903000.0  
0001193125-12-080059   4921000.0  
0001193125-12-080059  14577000.0

In [59]:
import numpy as np
np.reshape(indexing.value, (1,indexing.shape[0]*indexing.shape[1]))

indexing

ValueError: cannot reshape array of size 10 into shape (1,30)

In [39]:

indexing = indexing.stack()

In [40]:
indexing.to_frame()

0
adsh                 tag                                                ddate                     
0001376231-12-000002 RetainedEarningsAccumulatedDeficit                 20101231 value   -107439.0
                                                                        20111231 value   -112835.0
0001193125-12-080059 DueFromRelatedPartiesCurrent                       20111231 value   5903000.0
                                                                        20101231 value   4921000.0
                     IncreaseDecreaseInOtherAccountsPayable             20111231 value  14577000.0
                                                                        20101231 value  15888000.0
                                                                        20091231 value -51868000.0
                     OtherComprehensiveIncomeDerivativesQualifyingAs... 20091231 value    369000.0
                                                                        20101231 value    891000.0
                                                                        20111231 value   -519000.0

In [22]:
grouped = quarter_numbers.groupby('adsh')
grouped.groups

{'0000002178-12-000010': Int64Index([ 115416,  115417,  115418,  115419,  115420,  115421,  115422,
              115423,  115424,  115425,
             ...
             2342572, 2342573, 2342574, 2342575, 2342683, 2342684, 2342685,
             2342692, 2342693, 2342694],
            dtype='int64', length=255),
 '0000003453-12-000026': Int64Index([  73429,   73430,  115433,  115434,  115435,  115436,  115437,
              115438,  115439,  115440,
             ...
             3615100, 3615101, 3615102, 3615103, 3615106, 3615108, 3615109,
             3615114, 3615115, 3615116],
            dtype='int64', length=1953),
 '0000003499-12-000005': Int64Index([  73431,   73432,   73433,   73434,   73435,   73436,   73437,
               73438,   73439,   73440,
             ...
             3616469, 3616470, 3616471, 3617830, 3618131, 3618132, 3618133,
             3618134, 3618135, 3618136],
            dtype='int64', length=690),
 '0000003570-12-000023': Int64Index([  73443,   73444,   

In [24]:
type(grouped)

pandas.core.groupby.groupby.DataFrameGroupBy

In [29]:
grouped.groups['0000002178-12-000010']

Int64Index([ 115416,  115417,  115418,  115419,  115420,  115421,  115422,
             115423,  115424,  115425,
            ...
            2342572, 2342573, 2342574, 2342575, 2342683, 2342684, 2342685,
            2342692, 2342693, 2342694],
           dtype='int64', length=255)

In [21]:
pivot_test = grouped.pivot(index = 'adsh',columns = 'tag', values ='value')

AttributeError: Cannot access callable attribute 'pivot' of 'DataFrameGroupBy' objects, try using the 'apply' method

In [11]:
#Valeant'sx numbers
# BBS_10k = quarter_numbers.loc[quarter_numbers['adsh'] == '0001193125-12-117573']

# homex_USD = homex_fiings.loc[homex_fiings['uom']== 'USD']
# valeant_USD

In [12]:
# len(all_valeant)

In [13]:
# all_nums = list(set(BBS_10k.value.abs()))
# all_nums

In [14]:
# len(all_nums)

In [15]:
# def first_digit(number):
#     return int(str(number)[0])    

In [16]:
# all_nums_fd = [first_digit(num) for num in all_nums] 
# len(all_nums_fd)

In [17]:
# all_nums_fd

In [18]:
# def frequencies(first_digits):
#     counts = [0]*10
#     for x in first_digits:
#         if x > 0:
#             counts[x] += 1 
#     total = sum(counts)
#     freq = [count/total for count in counts]
# #     print(sum(freq))
#     return freq[1:] #same as going from element 1 through the end (i.e. 9)

In [19]:
# frequencies(all_nums_fd)

In [20]:
# import math
# r = range(10)
# benfords_law = [math.log10(1 + 1/digit) for digit in r if digit != 0]
# benfords_law

In [21]:
# import plotly

# plotly.offline.init_notebook_mode(connected=True)
# Benfords_Law = {'type': 'scatter', 'x': list(range(1, 10)), 'y': benfords_law}
# All_nums = {'type': 'scatter', 'x': list(range(1, 10)), 'y': frequencies(all_nums_fd)}


# plotly.offline.iplot([Benfords_Law, All_nums])

In [22]:
# import math
# def kl_divergence(freq):
#     kl_div = 0.0
#     for d in range(1, 10):
#         Q = (math.log(d+1) - math.log(d)) / math.log(10) #calculates theoretical benfords (perfect benny freqs)
#         P = freq[d-1]
#         kl_div += math.log(P / Q) * P
# #         print((freq))

#     return round(float(kl_div),3)

In [23]:
# kl_divergence(frequencies(all_nums_fd))

In [24]:
# kl_divergence(benfords_law)

In [25]:
# BBS_all = quarter_numbers.loc[quarter_numbers['entity_CIK'] == '0001193125']

In [26]:
# BBS_all.describe(include = 'all')

In [27]:
# check_cik= quarter_numbers.loc[quarter_numbers['entity_CIK'] == '0001193125']
# check_cik

In [28]:
# check_filing= quarter_numbers.loc[quarter_numbers['adsh'] == '0001193125-12-089232']
# check_filing #.tag['Goodwill']